### DISCUSSION TITLE
Router: match next route on non existing module/controller/action

### DISCUSSION THREAD

**Discussion by gianks** [phalcon/cphalcon#15664] [Q&A]
Hi,
when using placeholders for module, controller or action i'd expect that a rule should not match if the requested URI parts do not fit an existing combination of those.

Instead the router matches the pattern anyway and then forwards internally to the 404, despite the presence of a rule with less precedence which would match and work as expected.

Consider the following ruleset for example:
```

    $router->add("/{locale:[a-z]{2}}/{page_slug:[A-Za-z\-\_]+}/?",
            array(
                "locale"     => 1,
                "controller" => "pages",
                "action"     => "showPage",
          )
    );

    $router->add("/{locale:[a-z]{2}}/:controller/?",
            array(
                "locale"     => 1,
                "controller" => 2,
                "action"     => "index",
            )
    );

```
In this scenario the first rule will never match because of the second, which matches the same regex essentially.
Why is the router not moving to the next rule if the specified controller does not exist?
How can i make this work as expected?

Below i can show how i solved it to make it work in the meantime but it's not the same:

```

    $router->add("/{locale:[a-z]{2}}/pages/{page_slug:[A-Za-z\-\_]+}/?",
            array(
                "locale"     => 1,
                "controller" => "pages",
                "action"     => "showPage",
          )
    );

    $router->add("/{locale:[a-z]{2}}/:controller/?",
            array(
                "locale"     => 1,
                "controller" => 2,
                "action"     => "index",
            )
    );

```

**Comment by quasipickle** [comment]
That's just not how it works.  The router finds the first route that matches the URL.  Then the dispatcher discovers the route doesn't exist and does a 404.

Fortunately, the order in which routes are matched is defined - last first.  So you want to add your routes from most generic to most specific.  In this case, add your `pages` route _after_ your generic `controller` route.

**Reply by gianks** [reply]
Thanks but what you suggested won't work as desired anyway.
I know rules are applied bottom-up but this does not solve the problem of loading a page in place of a non-existing controller.

My first example won't load any "page" since the page_slug is always intercepted by the router as a :controller by the second rule.

Swapping (as you suggested and as i attempted before asking here) the two routes will do the opposite, each request will be sent to the pages controller and the appropriate controller will never be invoked even when exists.
And this makes perfect sense since the page_slug can't be evaluated nor by the router or the dispatcher, it's an arbitrary value.
But this is not true for the :controller placeholder, thus the dispatcher invokes correctly a 404, because it can determine the provided URI :controller doesn't exist.

To obtain the desired effect, the logic of the dispatcher shall be changed to attempt to match router's next rules before doing a 404 when the requested Module/Controller/Action is not found.
Maybe this can be achieved with the beforeMatch hook in the router, but i suspect this would not be the best way to do it, performance wise (since the controller lookup would be done twice).

Hope i made my case clear now.
Thanks

**Reply by quasipickle** [reply]
Can you modify your original post to include a problem URL and what the result is.

**Comment by wurst-hans** [comment]
You can try using a routing middleware. It should check the existence of slug and return `false` if it does not exist. So you can control which route is going to be used.
